src/plot.js

```diff
@@ -279,17 +327,57 @@ class Facet extends Mark {
       }
       marksChannels.push(markChannels);
     }
+    this.layout = function(index, scaleDescriptors, values, dimensions, options) {
+      const {marks, marksChannels, facetsKeys, marksIndexByFacet} = this;
+      const marksValues = this.marksValues = new Array(marks.length);
+      const scales = ScaleFunctions(scaleDescriptors);
+      const {fx, fy} = scales;
+      const fyMargins = fy && {marginTop: 0, marginBottom: 0, height: fy.bandwidth()};
+      const fxMargins = fx && {marginRight: 0, marginLeft: 0, width: fx.bandwidth()};
+      const subdimensions = {...dimensions, ...fxMargins, ...fyMargins};
+
+      // apply each mark’s layout
+      const newScaleChannels = new Map();
+      const layouts = [];
+      const newOptions = {};
+      for (let i = 0; i < marks.length; ++i) {
+        const mark = marks[i];
+        marksValues[i] = applyScales(marksChannels[i], scales);
+        if (true) {
+          const index = facetsKeys.map(key => marksIndexByFacet.get(key)[i]);
+          const {index: newIndex, values: newValues, channels} = layouts[i] = applyLayout.call(mark, mark.layout, index, scaleDescriptors, marksChannels[i], subdimensions, options);
+          marksValues[i] = {...values, ...newValues};
+          facetsKeys.forEach((key, j) => marksIndexByFacet.get(key)[i] = newIndex[j]);
+          if (channels) {
+            for (const [, {scale, options: hint}] of channels) {
+              if (scale) newOptions[scale] = {...hint, ...options[scale]};
+            }
+            addScaleChannels(newScaleChannels, channels, newOptions);
+            marksChannels[i] = channels;
+          }
+        }
+      }
+
+      const newScaleDescriptors = Scales(newScaleChannels, newOptions);
+      Object.assign(scaleDescriptors, newScaleDescriptors, scaleDescriptors);
+      const rescales = {...ScaleFunctions(newScaleDescriptors), ...scales};
+      for (let i = 0; i < marks.length; ++i) {
+        marksValues[i] = {...marksValues[i], ...layouts[i].values, ...applyScales(layouts[i].channels || [], rescales)};
+      }
```

Comment:
This code for facets is closely related to this non-faceted code, and it would be nice to unify them somehow:

https://github.com/observablehq/plot/blob/352d08670b8dfda49f60f5595d079a7b9808399c/src/plot.js#L117-L122

Reply:
Yes. This is the case for a lot of code, and I've wondered several times if we should not route everything through the faceted code, even when there is no faceting (the non-faceted case being just 1 facet with all the data, and null fx/fy). This might also help to clear up the difference between faceted and non-faceted indexes?

Reply:
Maybe! I tried before and the current state in main is the best I could come up with, but there’s always room for improvement. 😁 A little code duplication isn’t harmful, especially if it’s not visible externally, and there are usually ways to extract shared logic into shared helper functions to minimize disruption. This was just a reminder to look for a way to reduce the duplication, if possible.

(I’ve broken faceting in my working branch, but I’ll get back to it soon.)